### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by StasShilovich**

### CODE REVIEW

File: controller/src/main/java/com/epam/esm/controller/exception/RestExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                message + " " + exception.getLocalizedMessage()));
+    }
+
+    @ExceptionHandler(RuntimeException.class)
+    private ResponseEntity<ErrorResponse> handleNotExistException(RuntimeException exception) {
+        String internationale = messageSource.getMessage("runtime", null, locale);
+        String message = exception.getClass().getName() + " : " + exception.getMessage();
+        return ResponseEntity.status(INTERNAL_SERVER_ERROR).body(buildErrorResponse(INTERNAL_SERVER_ERROR,
+                internationale + " " + message));
+    }
```
Comment (TitenkovEA) [inline controller/src/main/java/com/epam/esm/controller/exception/RestExceptionHandler.java:52]: new line

File: controller/src/main/java/com/epam/esm/controller/exception/RestExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    @ExceptionHandler(IncorrectArgumentException.class)
+    private ResponseEntity<ErrorResponse> handleNotExistException(IncorrectArgumentException exception) {
```
Comment (TitenkovEA) [inline controller/src/main/java/com/epam/esm/controller/exception/RestExceptionHandler.java]: check the naming of your methods for exceptions handling, they should be appropriate

File: model/src/main/java/com/epam/esm/model/common/ParameterParser.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package com.epam.esm.model.common;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class ParameterParser {
```
Comment (TitenkovEA) [inline model/src/main/java/com/epam/esm/model/common/ParameterParser.java]: final

File: model/src/main/java/com/epam/esm/model/dao/entity/SortType.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package com.epam.esm.model.dao.entity;
+
+public enum SortType {
```
Comment (TitenkovEA) [inline model/src/main/java/com/epam/esm/model/dao/entity/SortType.java]: better move this to the common package

File: model/src/main/java/com/epam/esm/model/dao/DaoConstant.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package com.epam.esm.model.dao;
+
+public class DaoConstant {
```
Comment (TitenkovEA) [inline model/src/main/java/com/epam/esm/model/dao/DaoConstant.java]: final

File: model/src/main/java/com/epam/esm/model/service/impl/GiftCertificateServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    @Override
+    @Transactional
+    public List<CertificateDTO> filterByParameters(
+            String tag, String part, String sortBy, SortType type, int page, int size)
+            throws ServiceException, IncorrectArgumentException {
+        try {
+            long count = count();
+            Page certificatePage = new Page(page, size, count);
+            List<String> tags = ParameterParser.parseWithAndCondition(tag);
```
Comment (TitenkovEA) [inline model/src/main/java/com/epam/esm/model/service/impl/GiftCertificateServiceImpl.java]: I think you should not parse it in the service layer.
And I really believe that your controller will be able to parse this tags list for you (e.g. https://medium.com/@AADota/spring-passing-list-and-array-of-values-as-url-parameters-1ed9bbdf0cb2 )

File: model/src/main/java/com/epam/esm/model/service/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            return user.map(userMapper::toDTO)
+                    .orElseThrow(() -> new NotExistEntityException("User with id=" + id + " not exist!"));
+        } catch (DataAccessException e) {
+            log.error("Find user service exception", e);
+            throw new ServiceException("Find user service exception", e);
+        }
+    }
+
+    @Override
+    public List<UserDTO> findAll(int page, int size) throws ServiceException, IncorrectArgumentException {
```
Comment (TitenkovEA) [inline model/src/main/java/com/epam/esm/model/service/impl/UserServiceImpl.java:47]: what about the @Transactional
also why you do not use the @Transactional for count methods?

**Review by TitenkovEA — CHANGES_REQUESTED** [review]

File: controller/src/main/java/com/epam/esm/controller/exception/RestExceptionHandler.java (hunk 1 above)
Comment (StasShilovich) [reply controller/src/main/java/com/epam/esm/controller/exception/RestExceptionHandler.java:52]: fix

File: controller/src/main/java/com/epam/esm/controller/exception/RestExceptionHandler.java (hunk 2 above)
Comment (StasShilovich) [reply controller/src/main/java/com/epam/esm/controller/exception/RestExceptionHandler.java]: fix

File: model/src/main/java/com/epam/esm/model/common/ParameterParser.java (hunk 3 above)
Comment (StasShilovich) [reply model/src/main/java/com/epam/esm/model/common/ParameterParser.java]: fix

File: model/src/main/java/com/epam/esm/model/dao/entity/SortType.java (hunk 4 above)
Comment (StasShilovich) [reply model/src/main/java/com/epam/esm/model/dao/entity/SortType.java]: fix

File: model/src/main/java/com/epam/esm/model/dao/DaoConstant.java (hunk 5 above)
Comment (StasShilovich) [reply model/src/main/java/com/epam/esm/model/dao/DaoConstant.java]: fix

File: model/src/main/java/com/epam/esm/model/service/impl/GiftCertificateServiceImpl.java (hunk 6 above)
Comment (StasShilovich) [reply model/src/main/java/com/epam/esm/model/service/impl/GiftCertificateServiceImpl.java]: fix

File: model/src/main/java/com/epam/esm/model/service/impl/UserServiceImpl.java (hunk 7 above)
Comment (StasShilovich) [reply model/src/main/java/com/epam/esm/model/service/impl/UserServiceImpl.java:47]: fix

**Review by TitenkovEA — APPROVED** [review]

File: controller/src/main/java/com/epam/esm/controller/rest/GiftCertificateController.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    ) throws ServiceException, IncorrectArgumentException {
+        List<CertificateDTO> list = certificateService.filterByParameters(tags, part, sortBy, type, page, size);
+        long count = certificateService.count();
+        PagedModel.PageMetadata pageMetadata = new PagedModel.PageMetadata(size, page, count);
+        List<Link> linkList = buildLink(tags, part, sortBy, type, page, size, pageMetadata.getTotalPages());
+        PagedModel<CertificateDTO> pagedModel = PagedModel.of(list, pageMetadata, linkList);
+        return ResponseEntity.ok(pagedModel);
+    }
+
+    private List<Link> buildLink(List<String> tags, String part, String sortBy, SortType type, int page, int size,
```
Comment (SenPavel) [inline controller/src/main/java/com/epam/esm/controller/rest/GiftCertificateController.java:133]: Doesn't this method already exist in "CommonController" class?

File: model/src/main/java/com/epam/esm/model/dao/entity/User.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+public class User extends GenericEntity {
+    
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column
+    private String name;
+    @Column
+    private String surname;
+    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.LAZY, orphanRemoval = true)
```
Comment (SenPavel) [inline model/src/main/java/com/epam/esm/model/dao/entity/User.java]: No need to specify FetchType.Laze as it is lazy by default

File: model/src/main/java/com/epam/esm/model/dao/impl/GiftCertificateDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+            predicate = builder.equal(root.join(TAGS).get(NAME), tags.get(0));
+        } else {
+            Predicate tagOne = builder.equal(root.join(TAGS).get(NAME), tags.get(0));
+            Predicate tagTwo = builder.equal(root.join(TAGS).get(NAME), tags.get(1));
+            predicate = builder.and(tagOne, tagTwo);
+        }
+        return predicate;
+    }
+
+    private boolean buildUpdateQuery(CriteriaUpdate<GiftCertificate> update,
```
Comment (SenPavel) [inline model/src/main/java/com/epam/esm/model/dao/impl/GiftCertificateDaoImpl.java]: This method is too complicated, it sets params and defines if params were changed, move it to separate class and divide the logic  of "setting params" and "defining changed params"

File: model/src/main/java/com/epam/esm/model/service/GiftCertificateService.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    CertificateDTO find(Long id) throws ServiceException, NotExistEntityException;
+
+    CertificateDTO add(CertificateDTO certificateDTO) throws ServiceException;
+
+    CertificateDTO update(CertificateDTO certificateDTO) throws ServiceException;
+
+    void delete(Long id) throws ServiceException, NotExistEntityException;
+
+    List<CertificateDTO> filterByParameters(
+            List<String> tags, String part, String sortBy, SortType type, int page, int size)
```
Comment (SenPavel) [inline model/src/main/java/com/epam/esm/model/service/GiftCertificateService.java]: I suggest you to move these parameters to class as if you need to add new params for filtering, you will have to change interface and inherited classes

File: model/src/main/java/com/epam/esm/model/service/impl/GiftCertificateServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    private final GiftCertificateDao certificateDao;
+    private final GiftCertificateDTOMapper dtoMapper;
+
+    public GiftCertificateServiceImpl(GiftCertificateDao certificateDao, GiftCertificateDTOMapper dtoMapper) {
+        this.certificateDao = certificateDao;
+        this.dtoMapper = dtoMapper;
+    }
+
+    @Override
+    @Transactional
```
Comment (SenPavel) [inline model/src/main/java/com/epam/esm/model/service/impl/GiftCertificateServiceImpl.java]: If you want to annotate all methods as @Transcational you can put it at the class level

**Review by SenPavel — CHANGES_REQUESTED** [review]

File: model/src/main/java/com/epam/esm/model/service/impl/GiftCertificateServiceImpl.java (hunk 12 above)
Comment (StasShilovich) [reply model/src/main/java/com/epam/esm/model/service/impl/GiftCertificateServiceImpl.java]: fix

File: model/src/main/java/com/epam/esm/model/service/GiftCertificateService.java (hunk 11 above)
Comment (StasShilovich) [reply model/src/main/java/com/epam/esm/model/service/GiftCertificateService.java]: fix

File: model/src/main/java/com/epam/esm/model/dao/impl/GiftCertificateDaoImpl.java (hunk 10 above)
Comment (StasShilovich) [reply model/src/main/java/com/epam/esm/model/dao/impl/GiftCertificateDaoImpl.java]: fix

File: model/src/main/java/com/epam/esm/model/dao/entity/User.java (hunk 9 above)
Comment (StasShilovich) [reply model/src/main/java/com/epam/esm/model/dao/entity/User.java]: fix

File: controller/src/main/java/com/epam/esm/controller/rest/GiftCertificateController.java (hunk 8 above)
Comment (StasShilovich) [reply controller/src/main/java/com/epam/esm/controller/rest/GiftCertificateController.java:133]: "CommonController" used for controllers with "findAll" method, "GiftCertificateController" has "filterByParameter" method

**Review by SenPavel — APPROVED** [review]

**Pull request merged**